### DISCUSSION TITLE
Theia Backend Question

### DISCUSSION THREAD

**Discussion by bdresdner** [eclipse-theia/theia#13022] [Q&A] [answered]
I would like to be able to maintain different information for different users in my Theia environment. For example, when external APIs are invoked from my Theia extension's backend, different pieces of data need to be passed depending on the user.  When using Theia, do all users share the same backend?  I was thinking of maintaining a map with the userid as the key and the api info as the value.  Does it make sense to do this, or am I overengineering this?

**Comment by msujew** [comment]
> When using Theia, do all users share the same backend? I was thinking of maintaining a map with the userid as the key and the api info as the value. Does it make sense to do this, or am I overengineering this?

Theia is not a multi-tenant system. I.e. any Theia backend is only supposed to be used by a single user. Otherwise you get unwanted side effects such as shared plugins, settings and terminal processes.

Therefore, Theia backends need to be isolated from one another, you can do this via something like [theia cloud](https://theia-cloud.io/). With a kubernetes based solution, you will get a unique IP for each pod and you could differentiate them this way for an external API.

**Reply by bdresdner** [reply]
@msujew Thanks for the reply, and sorry for my confusion, on reading the Theia Cloud description, it looks like it supports multiple containers per pod and one user per container, correct?

**Reply by JonasHelming** [reply, marked as answer]
Yes, but Theia Cloud is a dedicated project with the goal to host Theia-based applications the way you described.

**Reply by bdresdner** [reply]
@JonasHelming Thank you.  I will look into using Theia Cloud.

**Reply by bdresdner** [reply]
@JonasHelming Just so I am clear on the purpose of Theia Cloud.  Is its primary purpose to help people integrate their Theia extensions into a Kubernetes/minikube environment?

**Reply by msujew** [reply]
@bdresdner It's primary purpose is to enable easy deployment of Theia based browser applications in the cloud using a Kubernetes engine - in a way that is well suited to Theia's single-tenancy design.

**Reply by bdresdner** [reply]
@msujew Thank you.  And it is not a hosting/subscription service?

**Reply by JonasHelming** [reply]
No, it is a very thin open source framework to be used to create your own hosting. Either for internal purposes or for you to create you own hosting service.

**Reply by msujew** [reply]
@bdresdner It's literally only an open source repository containing code that you point towards your own Kubernetes cluster.

**Reply by bdresdner** [reply]
@JonasHelming @msujew Thank you!

**Reply by goekce** [reply]
> Theia is not a multi-tenant system. I.e. any Theia backend is only supposed to be used by a single user. 

@msujew to clarify: It is not enough that every user starts their own backend on a multi-user OS environment but every backend must be started on a different OS environment like a container. Am I right?

**Reply by msujew** [reply]
Yes, at least it is designed that way - otherwise, users will have access to each other's workspace, which likely isn't what you actually want.

**Comment by goekce** [comment]
> users will have access to each other's workspace

Every user starts their own backend process in their home directory. It is not clear to me how another user would have then access to someone else's home directory.

**Reply by msujew** [reply]
I see now. Ok, that would make things into a single-tenant system again. I was under the assumption that users were using the same backend instance. Though, you obviously still have the issue of shared files, such as `/tmp`, something which you already noticed.

**Comment by koegel** [comment]
> > users will have access to each other's workspace
> 
> Every user starts their own backend process in their home directory. It is not clear to me how another user would have then access to someone else's home directory.

Sure, you can do that, but you need to create a solution that will create user accounts on the machine for each user, set up the backend for each user and make sure users are well isolated from each other and cannot access anything on the machine they should not be able to change. Also this will be difficult to scale. Doing this replicates what docker and kubernetes provide for you. It would be better imho to use docker and/or kubernetes and/or Theia Cloud.

**Reply by goekce** [reply]
I manage a [Jupyterhub](https://jupyterhub.readthedocs.io) for ~20 simultaneous users. Jupyterhub authenticates and isolates the users. Jupyterhub supports a process starter plugin called [`jupyter-server-proxy`](https://github.com/jupyterhub/jupyter-server-proxy) which [can be used  to start a Theia backend](https://github.com/jupyterhub/jupyter-server-proxy/tree/main/contrib/theia) for each user. Since I  have already authentication and autostart feature, I thought that installing Theia directly on my server would be a more lightweight and easier to manage solution.

However I cannot use this solution anyway:  It turned out that Theia's webview recently started [enforcing https on webviews]( https://github.com/eclipse-theia/theia/pull/6465#issuecomment-557494879) and `jupyter-server-proxy` only supports http requests. Probably I have to separate Theia from Jupyterhub by using kubernetes and/or Theia Cloud. So thanks @koegel :)
